### DISCUSSION TITLE
How to use vscode-nls and vscode-nls-dev to build a worker vscode extension?

### DISCUSSION THREAD

**Discussion by KrabsWong** [microsoft/vscode-discussions#146] [Extension Development QnA] [answered]
I've developing an extension (Web extension).

When I try to build i18n files with `vscode-nls-dev`, nothing files being built out.

Is there an example or more user guide? (The official doc is a little simple)

It's my gulp file below

```javascript
/* ---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
const del = require('del');
const gulp = require('gulp');
const sourcemaps = require('gulp-sourcemaps');
const ts = require('gulp-typescript');
const typescript = require('typescript');
const nls = require('vscode-nls-dev');

const tsProject = ts.createProject('./tsconfig.json', { typescript });
const outDest = 'out';

const languages = [
  { folderName: 'zh-cn', id: 'zh-cn' },
  { folderName: 'en', id: 'en' },
];

const cleanTask = function () {
  return del(['out/**', 'package.nls.*.json']);
};

const internalNlsCompileTask = function () {
  return tsProject.src()
    .pipe(sourcemaps.init())
    .pipe(tsProject()).js
    .pipe(nls.rewriteLocalizeCalls())
    .pipe(nls.createAdditionalLanguageFiles(languages, 'i18n', 'out'))
    .pipe(sourcemaps.write('../out', {
      // no inlined source
      includeContent: false,
      // Return relative source map root directories per file.
      sourceRoot: '../src',
    })).pipe(gulp.dest(outDest));
};

const addI18nTask = function () {
  return gulp.src(['package.nls.json'])
    .pipe(nls.createAdditionalLanguageFiles(languages, 'i18n'))
    .pipe(gulp.dest('.'));
};

const moveI18n = () => {
  return gulp
    .src(['./out/**/*.nls.*json'], { base: './out/' })
    .pipe(gulp.dest('./dist'));
};

gulp.task(
  'i18n',
  gulp.series(
    cleanTask,
    internalNlsCompileTask,
    addI18nTask,
    moveI18n
  )
);
```

I've create `package.nls.json` at the root folder, and build an folder named `i18n` 

![Tree of i18n folder](https://user-images.githubusercontent.com/12368943/193415749-48f8e69d-c87a-49ba-97fd-aba9b128558b.png)

**Comment by KrabsWong** [comment, marked as answer]
you can write the `package.nls.<languageCode>.json` manually for supporting the multiple language feature of pakcage.json file.

**Reply by isidorn** [reply]
fyi @TylerLeonhardt

**Reply by TylerLeonhardt** [reply]
Right now, we only support `package.nls.<language>.json` files for static strings in your package.json. There is currently an API proposal just added:
https://code.visualstudio.com/updates/v1_72#_localization-as-part-of-the-api

to bring localization into the API which will enable localization for web extensions.

**Reply by PeterWone** [reply]
A long-awaited change.

@KrabsWong - you can't use this in a published extension because it's a _proposed_ API. In the meantime, if you [look at my extension](https://github.com/PDConSec/vsc-print), I've been making this work for years in a way that you _can_ publish. 

In particular I direct your attention to https://github.com/PDConSec/vsc-print/blob/master/src/imports.ts which causes bundling to work properly, loads the resources and provides a function roughly equivalent to `t` from the proposed API. 

The proposed API does do one thing my solution can't. It supports token substitution. If you need this just embed $TOKENNAME into the strings and use `.replace(/$TOKENNAME/g, someValue)` to replace it.

**Reply by TylerLeonhardt** [reply]
We're hoping to finalize the API this month. Just an FYI

**Reply by KrabsWong** [reply]
> We're hoping to finalize the API this month. Just an FYI

It must be an amazing good news!

**Reply by KrabsWong** [reply]
> A long-awaited change.
> 
> @KrabsWong - you can't use this in a published extension because it's a _proposed_ API. In the meantime, if you [look at my extension](https://github.com/PDConSec/vsc-print), I've been making this work for years in a way that you _can_ publish.
> 
> In particular I direct your attention to https://github.com/PDConSec/vsc-print/blob/master/src/imports.ts which causes bundling to work properly, loads the resources and provides a function roughly equivalent to `t` from the proposed API.
> 
> The proposed API does do one thing my solution can't. It supports token substitution. If you need this just embed $TOKENNAME into the strings and use `.replace(/$TOKENNAME/g, someValue)` to replace it.

Many thanks bro, it's fantastic!
